### PR TITLE
avoid wrong handling for no-op network plugin

### PR DESCRIPTION
**Pull request by gongguan**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
/kind bug

**What this PR does / why we need it**:
When kubelet use no-op network, pod's nodeStatus equals nil is a normal condition (it's SetUpPod function done nothing) rather than an error.
If exception handling done as `TODO` or `FIXME`, no-op network may be handled as exception.
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/kubelet/dockershim/docker_sandbox.go [hunk 1]
```diff
@@ -337,6 +338,11 @@ func (ds *dockerService) getIPsFromPlugin(sandbox *dockertypes.ContainerJSON) ([
 		return nil, err
 	}
 	if networkStatus == nil {
+		if ds.network.PluginName() == network.DefaultPluginName {
+			//when docker use kubernetes.io/no-op, it will do nothing in SetUpPod function, networkStatus is nil
```
Comment (mattjmcnaughton) [inline pkg/kubelet/dockershim/docker_sandbox.go]: Can I suggest you rephrase this as:
```
// When docker uses the `DefaultPlugin` for networking, we expect the `networkStatus` to be nil, so we do not consider this an error.
```

File: pkg/kubelet/dockershim/docker_sandbox.go [hunk 2]
```diff
@@ -371,27 +377,31 @@ func (ds *dockerService) getIPs(podSandboxID string, sandbox *dockertypes.Contai
...
+	if err != nil {
+		// TODO: trusting the docker ip is not a great idea. However docker uses
+		// eth0 by default and so does CNI, so if we find a docker IP here, we
+		// conclude that the plugin must have failed setup, or forgotten its ip.
+		// This is not a sensible assumption for plugins across the board, but if
+		// a plugin doesn't want this behavior, it can throw an error.
+		// If error happens, warn but don't return an error, as pod status
+		// should generally not return anything except fatal errors
+		// FIXME: handle network errors by restarting the pod somehow?
+		klog.Errorf("failed to read pod IP from plugin/docker: %v", err)
```
Comment (mattjmcnaughton) [inline pkg/kubelet/dockershim/docker_sandbox.go]: Can you share more about why this changed from `klog.Warningf` to `klog.Errorf`?

File: pkg/kubelet/dockershim/docker_sandbox.go [hunk 3]
```diff
@@ -371,27 +377,31 @@ func (ds *dockerService) getIPs(podSandboxID string, sandbox *dockertypes.Contai
...
+	}
+
+	//while dockershim uses kubernetes.io/no-op network, no ip in networkStatus, pod will use docker ip
```
Comment (mattjmcnaughton) [inline pkg/kubelet/dockershim/docker_sandbox.go]: Can we rewrite this as:

```
When docker uses the `DefaultPlugin` for networking, the pod will use the docker ip.
```

**Review by mattjmcnaughton — COMMENTED** [review]
Thanks for your pr :)

Some suggestions from me around the comments.

One other question - what are the scenarios in which a cluster user would want to use the no-op network plugin? Am I correct in understanding that attempting to use the no-op network plugin should result in failure right now? Can we add some form of unit testing/integration testing to be more confident about this component of k8s?

/ok-to-test

**Comment by gongguan** [comment]
> what are the scenarios in which a cluster user would want to use the no-op network plugin? 
no-op network plugin may not used in cluster
>Am I correct in understanding that attempting to use the no-op network plugin should result in >failure right now?
use no-op network plugin won't fail, only throw a warning `failed to read pod IP from plugin/docker: Couldn't find network status for {pod name} through plugin: invalid network status for`. because `getIPs` function returns no error but ip list.
This patch only makes code more robust to avoid potential error. In current code, if error handled as TODO or FIXME, no-op network will be handled as an error
>Can we add some form of unit testing/integration testing to be more confident about this >component of k8s?
Forgive me being impolite, test to prove `networkStatus` is nil in no-op plugin? I don't have a good idea how to make the test neither abrupt nor complex yet.

**Comment by gongguan** [comment]
> what are the scenarios in which a cluster user would want to use the no-op network plugin? 
no-op network plugin may not used in cluster
>Am I correct in understanding that attempting to use the no-op network plugin should result in >failure right now?
use no-op network plugin won't fail, only throw a warning `failed to read pod IP from plugin/docker: Couldn't find network status for {pod name} through plugin: invalid network status for`. because `getIPs` function returns no error but ip list.
This patch only makes code more robust to avoid potential error. In current code, if error handled as TODO or FIXME, no-op network will be handled as an error
>Can we add some form of unit testing/integration testing to be more confident about this >component of k8s?
Forgive me being impolite, test to prove `networkStatus` is nil in no-op plugin? I don't have a good idea how to make the test neither abrupt nor complex yet.

File: pkg/kubelet/dockershim/docker_sandbox.go (hunk 1 above)
Comment (gongguan) [reply pkg/kubelet/dockershim/docker_sandbox.go]: done, thanks

File: pkg/kubelet/dockershim/docker_sandbox.go (hunk 3 above)
Comment (gongguan) [reply pkg/kubelet/dockershim/docker_sandbox.go]: done, thanks

File: pkg/kubelet/dockershim/docker_sandbox.go (hunk 2 above)
Comment (gongguan) [reply pkg/kubelet/dockershim/docker_sandbox.go]: If error exist here, it seems real failure happens, not no-op plugin's expected result

**Comment by mattjmcnaughton** [comment]
Going to defer to a different reviewer here :) I'm not sure I understand enough about the no-op network plugin to give useful feedback :)
